### PR TITLE
[BugFix] BE blocked in starrocks::KafkaDataConsumer::get_partition_offset

### LINKED ISSUE #8947: BE blocked in `starrocks::KafkaDataConsumer::get_partition_offset`
### Steps to reproduce the behavior (Required)
1. Create routine load
2. Stop all kafka brokers

### Expected behavior (Required)
The SR runs well.

### Real behavior (Required)
The SR would become slow. Many bthreads are blocked.
```
#0  0x00007f50ff200de2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x000000000613e9a9 in cnd_timedwait (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#2  0x000000000613ed53 in cnd_timedwait_abs (cnd=cnd@entry=0x214c76398, mtx=mtx@entry=0x214c76370, tspec=tspec@entry=0x7f4fb64ea910) at tinycthread_extra.c:106
#3  0x00000000060ec997 in rd_kafka_q_serve (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#4  0x00000000060b59ae in rd_kafka_query_watermark_offsets (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#5  0x0000000003afe143 in starrocks::KafkaDataConsumer::get_partition_offset (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#6  0x0000000003b0b91a in starrocks::RoutineLoadTaskExecutor::get_kafka_partition_offset (this=0xff8dc00, request=..., Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#7  0x0000000003b6b925 in starrocks::PInternalServiceImpl<doris::PBackendService>::get_info (this=0x1318a5e0, controller=<optimized out>, request=0x592b7bb90, response=0x1bdc2a740, done=<optimized out>) at /home/disk1/chaoyli/starrocks/be/../gensrc/build/gen_cpp/internal_service.pb.h:11970
#8  0x0000000004b5348e in brpc::policy::ProcessRpcRequest (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#9  0x0000000004b45727 in brpc::ProcessInputMessage (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#10 0x0000000004b465d3 in brpc::RunLastMessage::operator() (last_msg=0xba94200, this=<synthetic pointer>) at /home/disk2/wyb/workspace/wyb/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/brpc/input_messenger.cpp:141
#11 std::unique_ptr<brpc::InputMessageBase, brpc::RunLastMessage>::~unique_ptr (this=<synthetic pointer>, __in_chrg=<optimized out>) at /home/disk1/doris-deps/toolchain/installed/gcc-10.3.0/include/c++/10.3.0/bits/unique_ptr.h:361
#12 brpc::InputMessenger::OnNewMessages (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#13 0x0000000004bfba2e in brpc::Socket::ProcessEvent (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#14 0x0000000004a91a5f in bthread::TaskGroup::task_runner (Python Exception <type 'exceptions.NameError'> Installation error: gdb._execute_unwinders function is missing:
#15 0x0000000004a7ca21 in bthread_make_fcontext ()
#16 0x0000000000000000 in ?? ()
```

### StarRocks version (Required)
main@7ee482, 2.3-rc02, 2.2.3, 2.1.8

### PR DESCRIPTION
**Pull request by rickif**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #8947

## Problem Summary(Required) ：

`starrocks::KafkaDataConsumer::get_partition_offset` queries the partition offset from kafka.
The previous implementation would blocks many bthreads in `starrocks::KafkaDataConsumer::get_partition_offset` when all brokers of kafka are down. `starrocks::KafkaDataConsumer::get_partition_offset` would release bthread when the set timeout interval is reached.
This PR makes 2 attemp for the query. The first one is with 1s timeout, which would fail quickly when all brokers of kafka are down. The second one is with `routine_load_kafka_timeout_second` timeout, which would query more patiently when there's a real timeout.

### CODE REVIEW

File: be/src/runtime/routine_load/data_consumer.cpp [hunk 1]
```diff
@@ -253,13 +253,39 @@ Status KafkaDataConsumer::get_partition_offset(std::vector<int32_t>* partition_i
...
-                                                                      config::routine_load_kafka_timeout_second * 1000);
+
+        //When all kafka brokers are down,  query_watermark_offsets would block until timeout, which make the bthread unable to be preemptive.
+        // Here, we make 2 attempt to query_watermark_offsets.
+        // When all brokers are down, the 1st query_watermark_offsets with 1s timeout would fail quickly and return RdKafka::ERR__ALL_BROKERS_DOWN.
+        // When the 1st query_watermark_offsets fails but not due to all brokers are down, the 2nd query_watermark_offsets would try with timeout config::routine_load_kafka_timeout_second.
+
+        // fast path
+        RdKafka::ErrorCode err =
+                _k_consumer->query_watermark_offsets(_topic, p_id, &beginning_offset, &latest_offset, 1000 /*1000ms*/);
```
Comment (trueeyu) [inline be/src/runtime/routine_load/data_consumer.cpp:264]: The number of Bthreads of Brpc is limited, = the number of cores. If it is stuck for one second, will the problem also be serious

Comment (trueeyu) [reply be/src/runtime/routine_load/data_consumer.cpp:264]: Is it possible to change to asynchronous？

**Review by meegoo — COMMENTED** [review]
Is it possible direct access kafka from FE?

**Comment by rickif** [comment]
> 

@meegoo Of course we can, but this requires the network access from FE to Kafka brokers. IMO, it cannot be guranteed.

**Comment by meegoo** [comment]
> > 
> 
> @meegoo Of course we can, but this requires the network access from FE to Kafka brokers. IMO, it cannot be guranteed.

which scenario will it happen, since FE is our service entrance. And in the external table case, FE will also direct access the storage for meta.

**Comment by rickif** [comment]
> > > 
> > 
> > 
> > @meegoo Of course we can, but this requires the network access from FE to Kafka brokers. IMO, it cannot be guranteed.
> 
> which scenario will it happen, since FE is our service entrance. And in the external table case, FE will also direct access the storage for meta.

OK. I'm trying to move the check of offset to the FE.

**Comment by rickif** [comment]
> > > 
> > 
> > 
> > @meegoo Of course we can, but this requires the network access from FE to Kafka brokers. IMO, it cannot be guranteed.
> 
> which scenario will it happen, since FE is our service entrance. And in the external table case, FE will also direct access the storage for meta.

https://github.com/StarRocks/starrocks/pull/9085
@meegoo I've made the FE connect Kafka directly. PTAL.

**Pull request closed without merging**